wxWebView Edge (Chromium) Backend

TcT2k opened the pull request:
This adds an additional `wxWebView` backend based on the upcoming Microsoft Edge (Chromium). 

![image](https://user-images.githubusercontent.com/5075894/72071694-b0512500-32ec-11ea-9d74-0674293965aa.png)


The underlying WebView2 SDK is documented [here](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2).

Follow the build instructions in the [wxWebView documentation page](https://github.com/TcT2k/wxWidgets/blob/webview_edgec/interface/wx/webview.h#L282).

This is very similar to my previous attempt #807 where I integrated the (now legacy) Edge which is part of Windows 10.

This backend has currently quite a few features missing but might still be
a better choice than IE when displaying web content. The unsupported
features are the following: virtual filesystems, custom urls.

I've tested building with Visual Studio 2017 Community using CMake on Windows 10.
This is still a work in progress see the list of known todos:

- [x] Deciding on how to integrate
- [x] Document build process
- [x] Add to Visual Studio Projects
- [x] Check nmake builds
- [x] Add RunScript() implementation
- [x] Figure out string parsing from `ExecuteScript()`

TcT2k committed:
Add wxWebView::IsBackendAvailable()

This allows to check if a specified backend has been registered before creation of a new instance

TcT2k committed:
Added new build option wxUSE_WEBVIEW_EDGE

Commit:
Add wxWebView Edge (Chromium) implementation

This backend requires WebView2 SDK and enables usage of Edge (Chromium)
on Windows 7 and newer

TcT2k committed:
Load WebView2Loader.dll dynamically

TcT2k committed:
Add wxWebViewEdgeChromium to other build systems

TcT2k committed:
Add wxWebView methods to enable dev tools

Currently only implemented for the Edge (Chromium) backend.

TcT2k committed:
Rename from webview_edgec to webview_edge

TcT2k committed:
Set Edge data folder to UserLocalDataDir

TcT2k committed:
Implemented wxWebViewEdge::Print()

TcT2k committed:
Document wxWebViewEdge build process

PBfordev commented on include/wx/msw/webview_edge.h:
```diff
@@ -0,0 +1,186 @@
+// Licence:     wxWindows licence
+/////////////////////////////////////////////////////////////////////////////
+
+#ifndef wxWebViewEdgeC_H
```
Superminor, but: the "C" was removed from everywhere else, I would recommend removing it from the header guard as well.
```suggestion
#ifndef wxWebViewEdge_H
```

PBfordev commented on include/wx/msw/webview_edge.h:
```diff
@@ -0,0 +1,186 @@
+/////////////////////////////////////////////////////////////////////////////
+
+#ifndef wxWebViewEdgeC_H
+#define wxWebViewEdgeC_H
```
```suggestion
#define wxWebViewEdge_H
```

PBfordev commented on include/wx/msw/webview_edge.h:
```diff
@@ -0,0 +1,186 @@
+
+#endif // wxUSE_WEBVIEW && wxUSE_WEBVIEW_EDGE && defined(__WXMSW__)
+
+#endif // wxWebViewEdgeC_H
```
```suggestion
#endif // wxWebViewEdge_H
```

PBfordev reviewed (commented):
Thanks for doing that.

I just checked the build process and played with the sample a bit.

I have noticed that Edge webview failed to initialize with Edge Beta (`GetWebView2BrowserVersionInfo()` failed with _File not found_) which already had "Chrome" in the User Agent (...Chrome/79.0.3945.117 Safari/537.36 Edg/79.0.309.63). After installing Edge Dev (...Chrome/81.0.4003.0 Safari/537.36 Edg/81.0.381.0) the sample started working as expected. I could not display the Dev tools via local menu, `<Ctrl><Shift><i>` displayed a floating Dev window.

I suppose some of the missing functionality cannot be implemented via _RunScript()_ because of limitations on the returned strength length? For example, GetPageSource could be doctype+outerhtml?

**EDIT**
The above was tested with CMake build: VS 2019 Community, x64 DLL debug.

I now also tried the NMmake build: VS 2017 Express, x86 static release (_nmake -f makefile.vc BUILD=release USE_WEBVIEW_EDGE=1_).
I could not make it work, the generated _setup.h_ has `wxUSE_WEBVIEW_EDGE` set to 0. I could not find any mention of this define in _makefile.vc_ or _config.vc_.

TcT2k committed:
Update include guards

Co-Authored-By: PB <PBforDev@gmail.com>

TcT2k commented:
> 
> Thanks for doing that.
> 
Thanks for testing



> I just checked the build process and played with the sample a bit.
> 
> I have noticed that Edge webview failed to initialize with Edge Beta (`GetWebView2BrowserVersionInfo()` failed with _File not found_) which already had "Chrome" in the User Agent (...Chrome/79.0.3945.117 Safari/537.36 Edg/79.0.309.63). After installing Edge Dev (...Chrome/81.0.4003.0 Safari/537.36 Edg/81.0.381.0) the sample started working as expected. 

Each SDK release has a required minimum Edge Version as listed in the [release notes](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/releasenotes).

>I could not display the Dev tools via local menu, `<Ctrl><Shift><i>` displayed a floating Dev window.

Strange worked for me from the local menu (maybe it depends on what one clicks "Inspect"?)

> 
> I suppose some of the missing functionality cannot be implemented via _RunScript()_ because of limitations on the returned strength length? For example, GetPageSource could be doctype+outerhtml?

Could very well be implemented via RunScript(), but I currently didn't know how to. Some features can certainly be requested to be included in the WebView 2 SDK at their [feedback repo](https://github.com/MicrosoftEdge/WebViewFeedback/issues) (from what I've seen they are quite timely implementing reasonable feature requests).

> I now also tried the NMmake build: VS 2017 Express, x86 static release (_nmake -f makefile.vc BUILD=release USE_WEBVIEW_EDGE=1_).
> I could not make it work, the generated _setup.h_ has `wxUSE_WEBVIEW_EDGE` set to 0. I could not find any mention of this define in _makefile.vc_ or _config.vc_.

As far as I'm aware the MSW makefiles do not generate a _setup.h_ but just copy _setup0.h_ if necessary. So these settings have to manually be setup in _setup.h_.

PBfordev commented on interface/wx/webview.h:
```diff
@@ -279,6 +279,27 @@ class wxWebViewHandler
+    - Download the <a href="https://aka.ms/webviewnuget">WebView2 SDK</a>
+      nuget package (Version 0.8.355 or newer)
+    - Extract the package (it's a zip archive) to @c wxWidgets/3rdparty/webview2
+    - Enable @c wxUSE_WEBKIT_EDGE in CMake or @c setup.h
```
```suggestion
    - Enable @c wxUSE_WEBVIEW_EDGE in CMake or @c setup.h
```

PBfordev commented:
Thanks for the information. You are obviously right about setup.h and makefiles, it was just me being double-stupid.

TcT2k committed:
Update interface/wx/webview.h

Co-Authored-By: PB <PBforDev@gmail.com>

RobinD42 commented on interface/wx/webview.h:
```diff
@@ -279,6 +279,27 @@ class wxWebViewHandler
+    virtual filesystems, custom urls, text selection, find, source code.
+
+    This backend is not enabled by default, to build it follow these steps:
+    - Requires Visual Studio 2015 or newer
```
Possibly better wording: "Visual Studio 2015, or newer, is required."

RobinD42 commented on interface/wx/webview.h:
```diff
@@ -279,6 +279,27 @@ class wxWebViewHandler
+    - Build wxWidgets webview library
+    - Copy @c WebView2Loader.dll from @c wxWidgets/3rdparty/webview2/build/
+      to your applications executable
+    - At runtime your can use wxWebView::IsBackendAvailable() to check if the
```
"your can" --> "you can"

TcT2k committed:
Fix typo and wording in webview documentation

TcT2k commented:
@RobinD42 thanks for reviewing, I've included your suggestions.

hassandraga commented:
I will tray to build this tomorrow, and share results.

vadz commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+        userDataPath.wc_str(),
+        nullptr,
+        Callback<IWebView2CreateWebView2EnvironmentCompletedHandler>(
+            [this](HRESULT WXUNUSED(result), IWebView2Environment* environment) -> HRESULT
```
I'm not really against using C++11 in this file (as it can only be compiled with MSVS 2015 anyhow), but it's not clear to me that using lambdas improves readability of this code, especially for the longer functions below. Why not just make them normal functions/methods?

TcT2k replied:
This isn't really about lambda vs function. The `Microsoft::WRL::Callback<>` template implements an instance of event call back type (in this case `IWebView2CreateWebView2EnvironmentCompletedHandler`).
The template takes a lambda or a function pointer (which wouldn't work with a method pointer, if I'm not mistaken).
I think it improves readability immensely compared to the alternative where `wxWebViewEdge` would have to implement all these event interfaces.

From a pure technical point of view it could obviously be done, but I've no interest in doing it, especially not in this case where the underlying SDK recommends Visual Studio 2015 or newer.

vadz replied:
> This isn't really about lambda vs function. The `Microsoft::WRL::Callback<>` template implements an instance of event call back type (in this case `IWebView2CreateWebView2EnvironmentCompletedHandler`).
> The template takes a lambda or a function pointer (which wouldn't work with a method pointer, if I'm not mistaken).

It looks like `WRL::Callback` [can be constructed](https://docs.microsoft.com/en-us/cpp/cppcx/wrl/callback-function-wrl?view=vs-2019) from an object and its method, so I thought it would work. I could be wrong though, I've never used it. But my remark was really _just_ about lambdas vs functions.

TcT2k replied:
Didn't know that constructor (never saw it used in the samples), but with it readability can definitely be improved (especially for the longer lambdas).

vadz commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+{
+    wxVector<wxSharedPtr<wxWebViewHistoryItem> > backhist;
+    //As we don't have std::copy or an iterator constructor in the wxwidgets
+    //native vector we construct it by hand
```
This is not true (any more?) BTW and this code could perfectly well use `wxVector(first, last)` ctor.

TcT2k replied:
I would assume that is copy/paste from `webview_ie.cpp` (will take a look)

vadz commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+            return result;
+        else
+            return false;
+    */
```
Do we really need to keep this? If so, a comment explaining why is this code not used but still kept would be helpful.

vadz commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+
+wxString wxWebViewEdge::GetCurrentTitle() const
+{
+    PWSTR _title;
```
Use of underscore prefix is really unusual, why? Could we rename it to just `title` or does this clash with something (what?).

vadz reviewed (commented):
Thanks a lot once again!

Globally this is very good, of course, but there is this issue of `wxCOMPtr` (which we'll solve by finally making it public, I guess, unless someone objects) and also a couple of the usual naming/style nitpicks. Beyond this, I'd sprinkle the code with `wxLogApiError()` because this could be useful in case things don't behave as expected and ignoring error codes is usually not the best thing to do.

TcT2k committed:
Apply suggestions from code review

Co-Authored-By: VZ <vz-github@zeitlins.org>

TcT2k commented on include/wx/msw/webview_edge.h:
```diff
@@ -0,0 +1,186 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        include/wx/msw/webview_edge.h
+// Purpose:     wxMSW Edge Chromium wxWebView backend
+// Author:      Markus Pingel
```
He is a contributor in my company I've outsourced the initial work to.

TcT2k commented on include/wx/webview.h:
```diff
@@ -151,19 +152,22 @@ class WXDLLIMPEXP_WEBVIEW wxWebView : public wxControl
     virtual wxString GetPageText() const = 0;
     virtual bool IsBusy() const = 0;
     virtual bool IsContextMenuEnabled() const { return m_showMenu; }
+    virtual bool IsDevToolsEnabled() const { return false; }
```
I know it's strange, in the underlying API they use use `AreDevToolsEnabled` it's just that it's never used in wxWidgets and just seems strange in a different way to me :)
Currently I would prefer you `IsAccessToDevToolsEnabled()` suggestion.

TcT2k commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+
+bool wxWebViewEdge::Initialize()
+{
+#define RESOLVE_LOADER_FUNCTION(type, funcname)                        \
```
Copied that from other code, will look into `wxDL_INIT_FUNC()`

TcT2k commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+    LPWSTR versionStr;
+    if (SUCCEEDED(wxGetWebView2BrowserVersionInfo(NULL, &versionStr)))
+    {
+        if (versionStr)
```
[Documentation](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/webview2.idl#getwebview2browserversioninfo) doesn't specify it, but sure it seems to make sense to free it. Will look into that.

TcT2k replied:
Seems to be implied to call `CoTaskMemFree()` in COM APIs.

TcT2k committed:
Update setup.h files

TcT2k committed:
Use wxDL_INIT_FUNC() macro

TcT2k committed:
Remove explicit check for Windows 7

TcT2k commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+
+    // WebView2 is only available for Windows 7 or newer
+    if (!wxCheckOsVersion(6, 1))
+        return false;
```
My idea was blocking XP users, but sure if the WebView2Loader.dll was marked as Win7+ ist would fail loading and otherwise the `GetVersionInfo` call will fail. Will remove.

TcT2k committed:
Rename wxWebView::IsDevToolsEnabled() to IsAccessToDevToolsEnabled()

TcT2k commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+    }
+
+    if (!IsAvailable())
+        return false;
```
Make sense to move that

TcT2k commented on include/wx/msw/webview_edge.h:
```diff
@@ -0,0 +1,186 @@
+
+private:
+    bool m_initialized = false;
+    bool m_isBusy = false;
```
They were already initialized in code anyway, so could safely be removed.

TcT2k committed:
Various small code fixups

TcT2k committed:
Add a note about WebView2 SDK license requirements

TcT2k committed:
Add wxLogApiError() to important WebView2 API calls

TcT2k commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,777 @@
+
+    // Check if a Edge browser can be found by the loader DLL
+    LPWSTR versionStr;
+    if (SUCCEEDED(wxGetWebView2BrowserVersionInfo(NULL, &versionStr)))
```
Added to a few calls where the result might be helpful

TcT2k commented on include/wx/msw/webview_edge.h:
```diff
@@ -0,0 +1,186 @@
+#include "wx/control.h"
+#include "wx/dynlib.h"
+#include "wx/webview.h"
+#include "wx/msw/private/comptr.h"
```
Making `wxCOMPtr` public would probably be trivial. But it could also make sense to move the edge implementation to a private file so a using application wouldn't have it's include paths setup to include "Webview2.h".

vadz replied:
This is a good point that I didn't think about and would indeed be worth doing IMO.

TcT2k committed:
Move wxWebViewEdge implementation to private header

TcT2k committed:
Move WebView2 event handling from lambdas to functions

PBfordev commented on interface/wx/webview.h:
```diff
@@ -279,6 +279,29 @@ class wxWebViewHandler
+      backend can be used (it will be available if @c WebView2Loader.dll can be
+      loaded and Edge (Chromium) is installed)
+    - Make sure to add a note about using the WebView2 SDK to your application
+      documentation, as required by it's licence
```
```suggestion
      documentation, as required by its licence
```

TcT2k committed:
Call CoTaskMemFree() on LPWSTR return by WebView2 API

TcT2k committed:
Fix typo in interface/wx/webview.h

Co-Authored-By: PB <PBforDev@gmail.com>

TcT2k committed:
Remove outdated code and comment

TcT2k commented:
@vadz I think I've addressed all your suggestions, thanks for reviewing.

I would also like to thank the other reviewers, for feedback and suggested changes.

On a general not I just wanted to let everyone know that the stable version of edge released today is version _79_ which requires the older version _0.8.270_ of the WebView2 SDK. To use that older SDK one would have to comment out the `EnableContextMenus()` code.
It's not worth addressing this in code as the next stable version is only 6 weeks away at maximum.
TLDR
Use the **Dev Channel** for now.

MaartenBent commented on build/cmake/lib/webview/CMakeLists.txt:
```diff
@@ -34,6 +34,21 @@ endif()
     wx_lib_link_libraries(webview PUBLIC "-framework WebKit")
+elseif(WXMSW)
+    if(wxUSE_WEBVIEW_EDGE)
+        wx_lib_include_directories(webview PRIVATE ${PROJECT_SOURCE_DIR}/3rdparty/webview2/build/native/include)
```
In case the directory contains spaces.

```suggestion
        wx_lib_include_directories(webview PRIVATE "${PROJECT_SOURCE_DIR}/3rdparty/webview2/build/native/include")
```

MaartenBent commented on build/cmake/lib/webview/CMakeLists.txt:
```diff
@@ -34,6 +34,21 @@ endif()
+    if(wxUSE_WEBVIEW_EDGE)
+        wx_lib_include_directories(webview PRIVATE ${PROJECT_SOURCE_DIR}/3rdparty/webview2/build/native/include)
+
+        if (${CMAKE_SIZEOF_VOID_P} EQUAL 4)
```
```suggestion
        if (CMAKE_SIZEOF_VOID_P EQUAL 4)
```

MaartenBent commented on build/cmake/lib/webview/CMakeLists.txt:
```diff
@@ -34,6 +34,21 @@ endif()
+        endif()
+
+        add_custom_command(TARGET webview POST_BUILD
+            COMMAND  ${CMAKE_COMMAND} -E copy
```
```suggestion
            COMMAND ${CMAKE_COMMAND} -E copy
```

MaartenBent commented on build/cmake/lib/webview/CMakeLists.txt:
```diff
@@ -34,6 +34,21 @@ endif()
+        add_custom_command(TARGET webview POST_BUILD
+            COMMAND  ${CMAKE_COMMAND} -E copy
+            "${PROJECT_SOURCE_DIR}/3rdparty/webview2/build/native/${WEBVIEW2_ARCH}/WebView2Loader.dll"
+            $<TARGET_FILE_DIR:webview>/"WebView2Loader.dll")
```
```suggestion
            "$<TARGET_FILE_DIR:webview>/WebView2Loader.dll")
```

vadz commented on include/wx/webview.h:
```diff
@@ -151,19 +152,22 @@ class WXDLLIMPEXP_WEBVIEW wxWebView : public wxControl
     {
         m_showMenu = enable;
     }
+    virtual void EnableDevTools(bool WXUNUSED(enable) = true) { }
```
Oh, sorry for being annoying, but I forgot to mention that if we rename the accessor, we probably should rename this one to `EnableAccessToDevTools()` too, just for symmetry. It's not very important, of course, but IMHO it's still worth doing. If you'd like, I can do the renaming myself when merging.

MaartenBent commented on build/msw/makefile.gcc:
```diff
@@ -5261,109 +5266,76 @@ clean:
 setup_h: $(SETUPHDIR)\wx ..\..\include\wx\$(__SETUP_H_SUBDIR_FILENAMES)\setup.h $(SETUPHDIR)\wx\setup.h $(SETUPHDIR)\wx\msw\rcdefs.h
 
 $(LIBDIRNAME)\libwxregex$(WXUNICODEFLAG)$(WXDEBUGFLAG).a: $(WXREGEX_OBJECTS)
-	$(foreach f,$(subst \,/,$(WXREGEX_OBJECTS)),$(shell echo $f >> $(subst \,/,$@).rsp.tmp))
```
Lots of unrelated changes, because the bakefile version used to generate these files is newer than `0.2.11`, it now uses response files.

You could change the files of your bakefile installation with [these](https://github.com/vslavik/bakefile/commit/8e8bfbfd49331d1957f12ae549a655200fefb6a7) changes. Or maybe just let @vadz rebake before merging.

MaartenBent commented on build/msw/wx_webview.vcxproj:
```diff
@@ -138,7 +138,7 @@
     <ClCompile>
       <Optimization>Disabled</Optimization>
-      <AdditionalIncludeDirectories>$(OutDir)$(wxIncSubDir);..\..\include;..\..\src\tiff\libtiff;..\..\src\jpeg;..\..\src\png;..\..\src\zlib;..\..\src\regex;..\..\src\expat\lib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>$(OutDir)$(wxIncSubDir);..\..\include;..\..\src\tiff\libtiff;..\..\src\jpeg;..\..\src\png;..\..\src\zlib;..\..\src\regex;..\..\src\expat\lib;..\..\3rdparty\webview2\build\native\include</AdditionalIncludeDirectories>
```
Why remove `%(AdditionalIncludeDirectories)` here and below?

TcT2k replied:
Not an intended changed, reverted.

MaartenBent commented on include/wx/msw/private/webview_edge.h:
```diff
@@ -0,0 +1,67 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        include/wx/msw/webview_edge.h
```
```suggestion
// Name:        include/wx/msw/private/webview_edge.h
```

vadz commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,801 @@
+                output->assign(result);
+        }
+        else
+            wxLogError(_("RunScript failed: %.8x"), error);
```
AFAICS this method is supposed to fill `output` with the error when it fails, and it's logged by the code in the caller. So this should be something like

```suggestion
            output->Printf("%s (0x%08lx)", wxSysErrorMsgStr(error), error);
            return false;
```

MaartenBent commented on include/wx/msw/webviewhistoryitem_ie.h:
```diff
@@ -11,7 +11,7 @@
 #include "wx/setup.h"
 
-#if wxUSE_WEBVIEW && wxUSE_WEBVIEW_IE && defined(__WXMSW__)
+#if wxUSE_WEBVIEW &&  (wxUSE_WEBVIEW_IE || wxUSE_WEBVIEW_EDGE) && defined(__WXMSW__)
```
```suggestion
#if wxUSE_WEBVIEW && (wxUSE_WEBVIEW_IE || wxUSE_WEBVIEW_EDGE) && defined(__WXMSW__)
```

vadz reviewed (commented):
Thanks a lot for your fixes!

I'd just one more renaming and, unless I'm mistaken, the fix in `RunScriptSync()` and I think it's ready to be merged.

Thanks again!

MaartenBent commented on samples/ownerdrw/makefile.gcc:
```diff
@@ -218,10 +218,7 @@ clean:
 	-if exist $(OBJS)\ownerdrw.exe del $(OBJS)\ownerdrw.exe
 
 $(OBJS)\ownerdrw.exe: $(OWNERDRW_OBJECTS) $(OBJS)\ownerdrw_ownerdrw_rc.o
```
Unrelated change, same as some other sample makefiles below.

MaartenBent commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,801 @@
+
+wxWebViewZoom wxWebViewEdge::GetZoom() const
+{
+    double old_zoom_factor = 0.0f;
```
Why use `f` here and below when it is a double?

MaartenBent commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,801 @@
+        m_impl->m_webView->Stop();
+}
+
+void wxWebViewEdge::Reload(wxWebViewReloadFlags WXUNUSED(flags))
```
Document that this flag is not used?

MaartenBent commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,801 @@
+
+}
+
+void wxWebViewEdge::DoSetPage(const wxString& html, const wxString& WXUNUSED(baseUrl))
```
Document that the second argument is not used?

MaartenBent reviewed (commented):
I also tested it. The main functionality (showing webpage, javascript) all works good.

I added lots of small suggestions and comments while going through the code.

I have build errors when precompiled headers are disabled, these headers are missing:
```
#include "wx/log.h"
#include "wx/msw/private.h"
```

There are a lot of empty functions. Maybe add a comment/log-message/TODO to these functions? This makes it easy to see what still needs to be done and what can't be done.

In the webview sample, there are some parts of the code related to IE emulation level and guarded by `wxUSE_WEBVIEW_IE` . I think `&& !wxUSE_WEBVIEW_EDGE` should be added to these. Or, probably better, check if the used backend is IE.

vadz commented:
@MaartenBent Thanks for your review, but I indeed decided not to mention the unrelated makefile changes as I can rebake them easily using latest bakefile 0.2 version. The situation with the generated makefiles is really pretty sad, but it's our/my own fault, so I won't bother patch/PR submitters with this, so please feel free to disregard this @TcT2k

TcT2k committed:
Rename wxWebView::EnableDevTools() to EnableAccessToDevTools()

TcT2k committed:
Improve wxWebViewEdge::RunScriptSync() error handling

TcT2k committed:
Add TODOs for unimplemented wxWebViewEdge features

TcT2k committed:
Apply suggestions from code review

Co-Authored-By: Maarten <MaartenBent@users.noreply.github.com>

TcT2k committed:
Implement title changed event

TcT2k committed:
Change handlers not supported message to debug

TcT2k committed:
Remove float constants when used with double

TcT2k committed:
Document unsupported parameters

TcT2k committed:
Only show IE specific menus in webview sample when using IE

TcT2k committed:
Restore %(AdditionalIncludeDirectories) in wx_webview.vcxproj

TcT2k commented:
@MaartenBent 
> There are a lot of empty functions. Maybe add a comment/log-message/TODO to these functions? This makes it easy to see what still needs to be done and what can't be done.

Makes sense, added.

> In the webview sample, there are some parts of the code related to IE emulation level and guarded by `wxUSE_WEBVIEW_IE` . I think `&& !wxUSE_WEBVIEW_EDGE` should be added to these. Or, probably better, check if the used backend is IE.

I've added a runtime check.

@MaartenBent @vadz Integrated your suggestions (and ignored the unintended bakefile changes).

TcT2k committed:
Revert "Remove outdated code and comment"

This reverts commit f48770ee5a60ad4b7455f126bf7dae35a44cefe3.

The changes broke back history returned

TcT2k committed:
Send wxEVT_WEBVIEW_LOADED

TcT2k committed:
Send wxEVT_WEBVIEW_NAVIGATED after updating history

TcT2k committed:
Fix javascript return values

Return values where wrapped by quotes because of
wrapper code used for IE compatibility.

TcT2k committed:
Implement text selection in wxWebViewEdge

Text selection/copy/paste etc implemented via javascript

PBfordev commented on src/msw/webview_edge.cpp:
```diff
@@ -0,0 +1,866 @@
+}
+
+wxString wxWebViewEdge::GetPageSource() const
+{
```
Is there a length limit on a string returned by `RunScript()` or the underlying API? If there is not, perhaps the page source could be returned by running `document.documentElement.outerHTML`? Similarly, the page text could be obtained by `document.body.textContent`. But TBH, my knowledge of javascript is basically nil, so...

BTW, in the sample I have noticed that the IE backend's `GetPageSource()` does not include the document type. When displaying the page source with the embedded browser context menu, the document type is there. I am MSW-only so I cannot test how this behaves with other backends. If needed, perhaps the document type could be obtained by calling `XMLSerializer().serializeToString(document.doctype)` and prepending it to the outerHTML.

TcT2k replied:
Implemented with your suggestions (used `innerText` instead of `textContent`).
I don't think we have to care about doctype for now if it's not currently returned by the IE backend.

MaartenBent commented:
I did one more (CMake) build test with `wxUSE_WEBVIEW_IE` disabled and `wxUSE_WEBVIEW_EDGE` enabled.
It requires some additional fixes:

<details>
<summary>diff in here</summary>

```diff
 build/cmake/init.cmake | 4 ++--
 include/wx/chkconf.h   | 3 ++-
 src/common/webview.cpp | 2 ++
 3 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/build/cmake/init.cmake b/build/cmake/init.cmake
index a0cfe8c541..f57863deb2 100644
--- a/build/cmake/init.cmake
+++ b/build/cmake/init.cmake
@@ -406,8 +406,8 @@ if(wxUSE_GUI)
                 wx_option_force_value(wxUSE_WEBVIEW OFF)
             endif()
         elseif(WXMSW)
-            if(NOT wxUSE_WEBVIEW_IE)
-                message(WARNING "WebviewIE not found or enabled, wxWebview won't be available")
+            if(NOT wxUSE_WEBVIEW_IE AND NOT wxUSE_WEBVIEW_EDGE)
+                message(WARNING "WebviewIE and WebviewEdge not found or enabled, wxWebview won't be available")
                 wx_option_force_value(wxUSE_WEBVIEW OFF)
             endif()
         elseif(APPLE)
diff --git a/include/wx/chkconf.h b/include/wx/chkconf.h
index 47c9f9c489..8deeff5e44 100644
--- a/include/wx/chkconf.h
+++ b/include/wx/chkconf.h
@@ -2290,7 +2290,8 @@
 #   endif
 #endif /* wxUSE_TREELISTCTRL */
 
-#if wxUSE_WEBVIEW && !(wxUSE_WEBVIEW_WEBKIT || wxUSE_WEBVIEW_WEBKIT2 || wxUSE_WEBVIEW_IE)
+#if wxUSE_WEBVIEW && !(wxUSE_WEBVIEW_WEBKIT || wxUSE_WEBVIEW_WEBKIT2 ||
+                       wxUSE_WEBVIEW_IE || wxUSE_WEBVIEW_EDGE)
 #   ifdef wxABORT_ON_CONFIG_ERROR
 #       error "wxUSE_WEBVIEW requires at least one backend"
 #   else
diff --git a/src/common/webview.cpp b/src/common/webview.cpp
index e7cd9c158f..0a46c5fccd 100644
--- a/src/common/webview.cpp
+++ b/src/common/webview.cpp
@@ -107,9 +107,11 @@ wxStringWebViewFactoryMap::iterator wxWebView::FindFactory(const wxString &backe
 void wxWebView::InitFactoryMap()
 {
 #ifdef __WXMSW__
+#if wxUSE_WEBVIEW_IE
     if(m_factoryMap.find(wxWebViewBackendIE) == m_factoryMap.end())
         RegisterFactory(wxWebViewBackendIE, wxSharedPtr<wxWebViewFactory>
                                                    (new wxWebViewFactoryIE));
+#endif
 
 #if wxUSE_WEBVIEW_EDGE
     if (wxWebViewEdge::IsAvailable() &&
```

</details>

MaartenBent committed:
Enable building without wxUSE_WEBVIEW_IE

TcT2k commented:
> I did one more (CMake) build test with `wxUSE_WEBVIEW_IE` disabled and `wxUSE_WEBVIEW_EDGE` enabled.

Probably some exotic config for the foreseeable future, but obviously no harm in fixing it. Applied.

TcT2k committed:
Implement GetPageSource() and GetPageText()

TcT2k commented:
I've come across one remaining problem, maybe @MaartenBent or @vadz  can help.

The [ExecuteScript()](https://docs.microsoft.com/en-us/microsoft-edge/hosting/webview2/reference/iwebview2webview#executescript) method of `IWebView2` returns a JSON encoded string. This is fine for integers, objects, etc, but for strings it would require some parsing on the client side. Like a C style string it can contain `\n`, quotes, etc. I haven't found any code in the wxWidgets code base capable of doing that, but maybe somebody has an idea on the easiest source for such functionality.

vadz commented:
Unfortunately I don't see any good solution. Their docs recommend using WinRT `RuntimeClass_Windows_Data_Json_JsonObject`, which is available to Win32 applications, but requires MSW 10 and I'd like to keep support for 7.

Parsing JSON in general (i.e. arrays, objects, ...) is definitely non-trivial enough to use a library for it, but the question is which one -- there are a bunch of them, especially as we can use C++11 in this code. And many of them are header-only and easy to integrate. Of course, it doesn't matter that much which one we choose if we only use it internally, but OTOH if we do include a 3rd party JSON library anyhow, it seems a bit stupid not to let people use it too, it could be pretty useful with e.g. client-side HTTP too.

For parsing strings only it's also not completely unreasonable to do it manually, I think...

TcT2k commented:
@vadz
> Unfortunately I don't see any good solution. Their docs recommend using WinRT `RuntimeClass_Windows_Data_Json_JsonObject`, which is available to Win32 applications, but requires MSW 10 and I'd like to keep support for 7.
> 

I also think Win 7 compatibly is important.

> 
> For parsing strings only it's also not completely unreasonable to do it manually, I think...

We would "just" have to parse escape sequences allowed in JSON, but I haven't found a library that only does that.

I've now implemented correct parsing using https://github.com/nlohmann/json it's header only and makes it really easy to implement, but it surely seems like little overkill. Maybe there could be future uses for it within wxWidgets.

vadz commented:
I don't really understand what's wrong with `ParseStringLiteral()` from UTF-8 point of view? AFAICS it should work just fine if we first pass the original string received from Edge to `wxString::FromUTF8()`. The real problem I see in it is rather that it doesn't handle surrogates correctly, e.g. `"\uD834\uDD1E"` should be converted to a string containing the single `U+1D11E` character under Unix (where `sizeof(wchar_t)==4`). But this shouldn't be difficult to fix... IMHO we should do this, call it `wxJSON::DecodeString()`, put it into some private header, add a couple of tests and use it.

Also, if we care about efficiency, we should use buffers and not `wxString` here, but this could be improved later.

I won't be able to do this immediately, but if you'd like, I could try doing it slightly later.

TcT2k committed:
Decode strings returned by ExecuteScript()

ExecuteScript returns strings as fully
quoted JSON strings so they have to be decoded
to a binary string.

TcT2k commented:
@vadz
> 
> 
> I don't really understand what's wrong with `ParseStringLiteral()` from UTF-8 point of view? AFAICS it should work just fine if we first pass the original string received from Edge to `wxString::FromUTF8()`. The real problem I see in it is rather that it doesn't handle surrogates correctly, e.g. `"\uD834\uDD1E"` should be converted to a string containing the single `U+1D11E` character under Unix (where `sizeof(wchar_t)==4`). But this shouldn't be difficult to fix... IMHO we should do this, call it `wxJSON::DecodeString()`, put it into some private header, add a couple of tests and use it.
> 

I've implemented and named a method like you suggested. Not sure if cases where `wchar_t` is 4 bytes need to be addressed as for now this is only used on windows.

In my tests it handles everything correct, but as with any string parsing please check it and feel free to improve it how you see fit.

Correctly parsing strings definitely helped bringing down some missing functionality via javascript calls.

PBfordev commented:
I have just tested (after 26c82d4, relevant User Agent part "Chrome/80.0.3987.53 Safari/537.36 Edg/80.0.361.33") the results of RunScript in the webview sample, compared the Edge to the IE backend and found two discrepancies:

**Return undefined**
Edge backend: "null"
IE backend: "undefined"
Edge browser: "undefined"

**Return date**
Edge backend: "{}"
IE backend (changing emul. lev., notice doubled quotation marks): ""2017-10-08T21:30:40.000Z""
Edge browser: Sun Oct 08 2017 23:30:40 GMT+0200 (Central European Summer Time)
IE 11 browser:  "Sun Oct 08 2017 23:30:40 GMT+0200 (Central Europe Daylight Time)"
Firefox:  "Sun Oct 08 2017 23:30:40 GMT+0200 (Central European Summer Time)"

vadz commented on include/wx/private/json.h:
```diff
@@ -0,0 +1,89 @@
+                if (wxIsxdigit(ch[1]) && wxIsxdigit(ch[2]) &&
+                    wxIsxdigit(ch[3]) && wxIsxdigit(ch[4]))
+                {
+                    wxUChar uchar = wxHexToDec(wxString(&ch[3], 2)) |
```
I'd use `wchar_t` explicitly here for clarity and add a comment saying that it works only when `sizeof(wchar_t) == 2`. In fact, it's probably better to write it like this:
```cpp
#if SIZEOF_WCHAR_T == 2
    // In this case, we handle surrogates without doing anything special was wchar_t strings use UTF-17 encoding.
    ... the existing code ...
#else 
     #error Implement correct surrogate handling.
#endif 
```

vadz reviewed (commented):
I'm ready to merge this, but there are a couple of changes that could/should be done for better JSON conformance.

BTW, we don't need nlohman JSON header any more, do we? GitHub still seems to show it as added...

TcT2k committed:
Improved JSON string parsing

TcT2k commented:
@vadz

Implemented your JSON parsing suggestions. It's indeed a little confusing that it's different what's valid in JSON and in javascript.

> 
> BTW, we don't need nlohman JSON header any more, do we? GitHub still seems to show it as added...

That header is not used anymore, not sure why it shows up for you in github (maybe a caching issue).

TcT2k commented:
@PBfordev 
> 
> 
> I have just tested (after [26c82d4](https://github.com/wxWidgets/wxWidgets/commit/26c82d43d180613909cb5a5a942d23379046c460), relevant User Agent part "Chrome/80.0.3987.53 Safari/537.36 Edg/80.0.361.33") the results of RunScript in the webview sample, compared the Edge and IE backends and found two discrepancies:
> 
I also see that discrepancies, but I'm pretty sure they are an issue in the underlying API. I've reported that upstream here:
https://github.com/MicrosoftEdge/WebViewFeedback/issues/105

hassandraga commented:
```
In file included from ..\..\include/wx/msw/private/webview_edge.h:16:0,
                 from ../../src/msw/webview_edge.cpp:29:
..\..\3rdparty\webview2\build\native\include/Webview2.h:422:24: fatal error: EventToken.h: No such file or directory
compilation terminated.
makefile.gcc:16197: recipe for target 'gcc_mswu\webviewlib_webview_edge.o' failed
mingw32-make: *** [gcc_mswu\webviewlib_webview_edge.o] Error 1
```
**Compiler:** TDM-GCC-32bit (SJLJ).
**VC Installed:** Visual Studio 2019 (with Windows 10 SDK).
**WebView2 SDK:** 0.8.355.

MaartenBent commented:
Same with other MinGW compilers. They have their own `w32api` implementation, that currently does not include the `winrt`/`wrl` headers.
Maybe the build system should check for MSVC compiler?

vadz commented:
I'd like to make this work with MinGW, but I'm not sure if it's doable with the current w32api. Even if it is, we can, and should, do it later, after merging this PR. For now we could maybe add a compiler check to `wx/chkconf.h` or maybe just mention in the comment in `setup.h` that this feature is only supported when using MSVC.

So should we merge this now (considering that the JSON inconsistency issues can't unfortunately be fixed) or is there anything else still remaining to fix? I don't see anything, but I could be missing something.

TcT2k commented:
I think this can be merged now.

vadz commented:
Done now, thanks again!

hassandraga commented:
Now it's time for supporting MinGW, I don't know if it's possible with MinGW Win32 API's.
